### PR TITLE
Flesh out DDL parsing functionality.

### PR DESCRIPTION
**Pull request by petermattis**
Added support for "SHOW DATABASES", "SHOW TABLES FROM" and "CREATE
UNIQUE INDEX".

Started work on CREATE TABLE. This is very much a work in progress, but
there should now be enough structure to create a proto.TableSchema from
the AST.

### CODE REVIEW

**Comment by tamird** [comment]
The majority of this is generated code, correct? Which part is which?

**Comment by petermattis** [comment]
`sql.go` is generated from `sql.y`.

File: sql/parser/ast.go [hunk 1]
```diff
@@ -37,20 +38,22 @@ type Statement interface {
...
+func (*Select) statement()         {}
+func (*Insert) statement()         {}
+func (*Update) statement()         {}
+func (*Delete) statement()         {}
+func (*Set) statement()            {}
+func (*Use) statement()            {}
+func (*CreateDatabase) statement() {}
+func (*CreateIndex) statement()    {}
+func (*CreateTable) statement()    {}
+func (*DDL) statement()            {}
```
Comment (vivekmenezes) [inline sql/parser/ast.go]: wanna put this in alphabetic order?

File: sql/parser/ast.go [hunk 2]
```diff
@@ -184,19 +187,119 @@ func (node *Use) String() string {
...
+	}
+	return buf.String()
+}
+
+// ColumnTableDef represents a column definition within a CREATE TABLE
+// statement.
+type ColumnTableDef struct {
+	Name       string
+	Type       ColumnType
+	Null       string // TODO(pmattis): Make this a bool.
```
Comment (vivekmenezes) [inline sql/parser/ast.go]: did you mean IsNullable?

File: sql/parser/ast.go [hunk 3]
```diff
@@ -205,22 +308,33 @@ const (
...
-	case astCreateIndex, astDropIndex:
+	case astDropIndex:
 		return fmt.Sprintf("%s %s ON %s", node.Action, node.Name, node.NewName)
 	case astRenameTable:
 		return fmt.Sprintf("%s %s %s", node.Action, node.Name, node.NewName)
-	case astCreateDatabase, astCreateTable, astCreateView:
-		return fmt.Sprintf("%s %s", node.Action, node.NewName)
+	case astCreateView:
+		return fmt.Sprintf("%s%s %s", node.Action, node.IfNotExists, node.NewName)
+	case astShowDatabases:
```
Comment (vivekmenezes) [inline sql/parser/ast.go]: What does show databases mean? Are we still going with Namespaces with Tables?

File: sql/parser/types.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+		fmt.Fprintf(&buf, "(%d)", node.N)
+	}
+	return buf.String()
+}
+
+// IntType represents an INT, TINYINT, SMALLINT, MEDIUMINT, BIGINT or INTEGER
+// type.
+type IntType struct {
+	Name     string
+	N        int
```
Comment (vivekmenezes) [inline sql/parser/types.go]: Is it worth making all these int types int64 to be more specific about their sizes?

**Comment by vivekmenezes** [comment]
LGTM

File: sql/parser/ast.go (hunk 1 above)
Comment (petermattis) [inline sql/parser/ast.go]: I believe these are in the order that they appear in the file. I'd prefer not to edit the order right now, though my OCD tends to agree with you.

File: sql/parser/ast.go (hunk 2 above)
Comment (petermattis) [inline sql/parser/ast.go]: Sure. That's a better name.

File: sql/parser/ast.go (hunk 3 above)
Comment (petermattis) [inline sql/parser/ast.go]: The vitess sql parser parses the MySQL variant of SQL. `SHOW DATABASES` comes from https://dev.mysql.com/doc/refman/5.1/en/show-databases.html. Namespace is not a term that is present in SQL. I'm not settled on this terminology, but it seems good to stick with the existing language for now.

File: sql/parser/types.go (hunk 4 above)
Comment (petermattis) [inline sql/parser/types.go]: `N` is the formatting width, not the value being stored. `IntType` is a descriptor for the type, not a holder of a particular value.

File: sql/parser/ast.go (hunk 3 above)
Comment (jusongchen) [inline sql/parser/ast.go]: Hi,

Does cockroachDB have a concept of DATABASE?  if not, having a command like
SHOW DATABASES could be misleading.

The term DATABASE  is used very differently by various RDBMS vendors (e.g.
a DATABASE in context of Oracle is completely different to that in MS SQL
Server) hence ANSI SQL avoids this term and uses the following schema terms:


4.2.9 Catalogs and schemas
4.2.9.1 Catalogs
A catalog is a named collection of SQL-schemas, foreign server descriptors,
and foreign data wrapper
descriptors in an SQL-environment. The mechanisms for creating and
destroying catalogs are implementationdefined.


4.2.9.2 SQL-schemas
An SQL-schema, often referred to simply as a schema, is a persistent, named
collection of descriptors. Any
object whose descriptor is in some SQL-schema is known as an SQL-schema
object.
A schema, the schema objects in it, and the SQL-data described by them are
said to be owned by the authorization
identifier associated with the schema.
SQL-schemas are created and destroyed by execution of SQL-schema statements
(or by implementation-defined
mechanisms).


4.2.12 SQL-data
SQL-data is data described by SQL-schemas — data that is under the control
of an SQL-implementation in an
SQL-environment.

4.3 Tables
A table is a collection of zero or more rows where each row is a sequence
of one or more column values. In
general, rows in a table are unordered; however, rows in a table are
ordered if the table is the result of a <query
expression> that immediately contains an <order by clause>. Each column has
a name and a data type. Each
row has exactly one value for each column C, and the data type of that
value is the data type of C.


thank you,

Jusong Chen


On Mon, Jun 22, 2015 at 7:00 AM, Peter Mattis <notifications@github.com>
wrote:

> In sql/parser/ast.go
> <https://github.com/cockroachdb/cockroach/pull/1430#discussion_r32936338>:
>
> >  		return fmt.Sprintf("%s %s ON %s", node.Action, node.Name, node.NewName)
> >  	case astRenameTable:
> >  		return fmt.Sprintf("%s %s %s", node.Action, node.Name, node.NewName)
> > -	case astCreateDatabase, astCreateTable, astCreateView:
> > -		return fmt.Sprintf("%s %s", node.Action, node.NewName)
> > +	case astCreateView:
> > +		return fmt.Sprintf("%s%s %s", node.Action, node.IfNotExists, node.NewName)
> > +	case astShowDatabases:
>
> The vitess sql parser parses the MySQL variant of SQL. SHOW DATABASES
> comes from https://dev.mysql.com/doc/refman/5.1/en/show-databases.html.
> Namespace is not a term that is present in SQL. I'm not settled on this
> terminology, but it seems good to stick with the existing language for now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1430/files#r32936338>.
>



-- 
Jusong Chen
408 315 2288

Comment (petermattis) [inline sql/parser/ast.go]: @jusongchen Cockroach will have the concept of a group of tables. Whether that is called a database, schema, catalog, namespace, directory or something else is still up in the air. The term database certainly has different meanings between DBs. Catalog feels like an obscure term to me. Directory is compelling due to the similarity to a filesystem directory.

Comment (jusongchen) [inline sql/parser/ast.go]: @PeterMattis Both Oracle and SQL server use term *schema *to name a
collection of schema objects.  An full qualifier of an Oracle schema object
has form of *schema.object_name@db_link *while SQL server one is
*server_name.database_name.schema_name.object_name.*

A Oracle schema has multiple namespaces within it while one SQL server
schema has only one. The following code runs well on Oracle. Please note a
table , an index and a constraint are all named 'person' and they are
created in one schema.


*create table person ( *
*psn_ID varchar(200) constraint pk_person primary key ,*
*first_name varchar(200),*
*last_name varchar(200) not null,*
*gender varchar(8)*
*)*
*;*

*create index person on person(last_name);*
*alter table person add constraint person check (gender in ('M','F',null));*

*alter table person drop constraint person;*

*drop index person;*
*drop table person;*



The reason I mention the above is that data migrations
between heterogeneous systems are always challenging.  A design decision
 which make data migration easier would greatly encourages customers to
adopt cockroach in the future.

Thank you,

Jusong Chen




On Mon, Jun 22, 2015 at 11:58 AM, Peter Mattis <notifications@github.com>
wrote:

> In sql/parser/ast.go
> <https://github.com/cockroachdb/cockroach/pull/1430#discussion_r32970342>:
>
> >  		return fmt.Sprintf("%s %s ON %s", node.Action, node.Name, node.NewName)
> >  	case astRenameTable:
> >  		return fmt.Sprintf("%s %s %s", node.Action, node.Name, node.NewName)
> > -	case astCreateDatabase, astCreateTable, astCreateView:
> > -		return fmt.Sprintf("%s %s", node.Action, node.NewName)
> > +	case astCreateView:
> > +		return fmt.Sprintf("%s%s %s", node.Action, node.IfNotExists, node.NewName)
> > +	case astShowDatabases:
>
> @jusongchen <https://github.com/jusongchen> Cockroach will have the
> concept of a group of tables. Whether that is called a database, schema,
> catalog, namespace, directory or something else is still up in the air. The
> term database certainly has different meanings between DBs. Catalog feels
> like an obscure term to me. Directory is compelling due to the similarity
> to a filesystem directory.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1430/files#r32970342>.
>



-- 
Jusong Chen
408 315 2288

**Pull request merged**